### ISSUE TITLE
Search engine icons for high-density displays

### ISSUE THREAD

**Issue by bijancamp** [Pitmairen/selection-search#37]
Hello again!

I recently started using a laptop with a high-pixel-density display (MacBook Pro, 15-inch, 2015), and the search engine icons I had previously been using now appear fuzzy on the new display. I tried using larger custom icons (greater than 16x16 pixels), but that did not seem to improve the rendering quality in the popup.

Here is a screenshot of the icons on the display:

<img width="336" alt="Selection Search Icons from 16x16 Canvas" src="https://user-images.githubusercontent.com/8800015/27518347-6ae29d0c-59a3-11e7-826c-5ce54a5798f1.png">

In case it's helpful, below are the sizes and URLs of the icons I am using. (I didn't include the icon for YouTube in the list, since I have to load it locally for now.) The listed size of each ICO file is that of the highest size in the file.

| Search Engine  | Icon Size | Icon URL |
| ------------- | ------------- | ------------- |
| Google  | 32x32  |  https://www.google.com/images/branding/product/ico/googleg_lodp.ico  |
| Wikipedia  | 48x48  |  https://www.wikipedia.org/static/favicon/wikipedia.ico  |
| Merriam-Webster  | N/A (SVG)  |  https://upload.wikimedia.org/wikipedia/commons/3/32/Merriam-Webster_logo.svg  |
| Amazon.com  | 48x48  |  https://www.amazon.com/favicon.ico  |

As a temporary solution, I changed the dimensions of the canvas and drawn image in [iconloader.js](https://github.com/Pitmairen/selection-search/blob/aba4c0a34badf3b98038f8a93b0c8524063bbf99/background/iconloader.js) from 16x16 to 32x32, which resulted in a much clearer rendering:

<img width="336" alt="Selection Search Icons from 32x32 Canvas" src="https://user-images.githubusercontent.com/8800015/27518707-9a6cb5a0-59ab-11e7-846d-8568149c1f34.png">
 
Do you it would be a good idea to make the size of the canvas and drawn image customizable? If so, should the size of the canvas be customizable per icon, or only for all icons as a whole (say, via Popup Menu -> Advance Settings or the CONFIG section for a custom style)?

I would be very happy to add a setting for this if you think it's a good idea. 😀

**Comment by Pitmairen** [comment]
Hi!

If the icons look good enough by changing the sizes in  iconloader.js to 32 then it may be good enough to just do this change without any configuration option. It shouldn't change anything other than maybe use a little bit more memory. So this would be the easiest way to fix it.

But if a config option is necessary, I think it would easiest with a global option in Popup Menu -> Advance Settings.

**Comment by bijancamp** [comment]
By changing the sizes in iconloader.js to 32 without any configuration option, do you mean making this change in the main branch so it will be the default for others?

In that case, do you know if changing the sizes to 32 in iconloader.js will decrease the quality of rendering 16x16 icons? It seems a lot of the icons that are generated from the getFaviconUrl method are 16x16.... I can test the effect on the quality of 16x16 icons when the sizes are increased on another machine I have. 😀

Yeah, I wasn't sure whether it should be a config option or not, since it seems some sources such as [this MSDN article](https://msdn.microsoft.com/library/gg491740(v=vs.85).aspx) give recommended sizes all the way up to 48x48. So I thought having a configurable setting would be good for that reason, but I definitely agree that fixing this without adding an option would be the easiest way. 😀

**Comment by Pitmairen** [comment]
Yes, I mean changing it so it becomes the default for everyone. I don't think this will decrease the quality of the 16x16 icons, at least not something that is noticeable. But it should be tested on a few icon to verify it. 

It would also have been nice if there was another service similar to the one currently used to get the favicons ( https://plus.google.com/_/favicon?domain=github.com), that would return icons with higher quality. But I haven't seen any other services like that, so currenty high quality icons must be added manually.

**Comment by Pitmairen** [comment]
This seems promising: https://github.com/mat/besticon

Maybe I can host this myself and use this to get better icons.

**Comment by bijancamp** [comment]
I just tested the effects of increasing the sizes to 32x32 in iconloader.js on a few 16x16 icons (on a lower density resolution display), and it seems that increasing the sizes results in the 16x16 icons being blurry. I find the blur noticeable, but I am not sure how many other people will notice.

Here are the icons as they appear with the current release (16x16 canvas):

![Selection Search Icons with 16x16 Canvas](https://user-images.githubusercontent.com/8800015/27567807-ca6bc526-5ab3-11e7-9ad3-c2f35f537bb2.png)

And here they are after increasing the canvas to 32x32:

![Selection Search Icons with 32x32 Canvas](https://user-images.githubusercontent.com/8800015/27567819-d8b62248-5ab3-11e7-8a92-122b923f1fdf.png)

Instead of hardcoding the sizes into iconloader.js, what do you think of automatically calculating the sizes based on the sizes of the icons? Basically, the sizes would be the same for every icon and would equal the size of the smallest icon. For example, if 4 icons are to be used, one with a size of 16x16 and three with a size of 32x32, then the sizes in iconloader.js would be set to 16x16 for all 4 icons. That way, the quality may not degrade in a noticeable way if smaller icons are used... although the quality would degrade if lower-sized icons are viewed on a higher density display....

besticon looks very interesting.... If that was used, maybe it would then be good to store the icons so they persist in between browser restarts, so that they don't have to be fetched too frequently from the server. Although local storage is limited, as you have mentioned in the issue on caching icons....

**Comment by Pitmairen** [comment]
Yes, they definitely look blurry in your screenshot, I did not see the same on my computer. There is a setting on the canvas that can be added to remove some of the blur, but I think it will make some other icons look worse. 
```
IconLoader._context.imageSmoothingEnabled = false;
```

So it is probably best not to hardcode the size. I think the size of the canvas can be changed dynamically, so we don't have to have the same size for every icon. I think adding the following to getDataURL in the IconLoader should work:
```
        var size = _img.naturalWidth;
        size = Math.max(16, size); // Minimum 16px
        size = Math.min(width, size); // Maximum 48px
        IconLoader._canvas.width = size;
        IconLoader._canvas.height = size;

        IconLoader._context.clearRect(0, 0, size, size);
        IconLoader._context.drawImage(_img, 0, 0, size, size);
        _urlCache = IconLoader._canvas.toDataURL("image/png");
        return _urlCache;
``` 

Maybe you can try this and see how the looks for you.

**Comment by bijancamp** [comment]
Thank you very much for those suggestions! I briefly tested the new logic for getDataURL on my high-density display, and it seemed to make the icons slightly more blurry for those that are greater than 48x48. (The 32x32 icons looked sharp though.) Here is how the icons looked before, whenever the sizes for all icons were hardcoded to 32x32:

<img width="202" alt="with 32 hardcoded" src="https://user-images.githubusercontent.com/8800015/27618680-61247e3e-5b83-11e7-8bac-1c1b744475fc.png">

And here is how they looked after applying the dynamic changes to getDataURL:

<img width="202" alt="with calculated sizes" src="https://user-images.githubusercontent.com/8800015/27618683-63cd492c-5b83-11e7-9ba2-17c1ed1a277d.png">

I can only see blurriness for the Wikipedia, Merriam-Webster, and Amazon.com icons (the ones that are 48x48 or larger.) Interestingly, the Merriam-Webster SVG icon, which has a naturalWidth of 300 pixels, seems to be the most distorted of the three....

Also, when testing, I changed the reference to "width" in the Math.min call to 48, since that is what I thought was probably intended....

I haven't gotten a chance to test the imageSmoothingEnabled property, or test on a lower-density display, but I will try to do more testing tomorrow, as well as play with the dynamic logic in getDataURL to maybe increase the sharpness. 😀

**Comment by Pitmairen** [comment]
It is weird that it gets more blurry with the dynamic size. What if you set the maximum size to 32px with the dynamic size? Is it different from when 32 is hard coded for all icons?

**Comment by bijancamp** [comment]
Yes, setting the maximum size to 32px with the dynamic size (using the icons listed in the first post) produces the exact same quality as when 32 is hard coded for all icons. So maybe the high-density display (220 pixels per inch) will blur images drawn on all canvas sizes larger than 32x32? Maybe each display has an optimal maximum size?

My lower density display (128 ppi) also seemed to have an optimal maximum size (at 16 pixels), as least when using 16x16 icons. 😀

**Comment by Pitmairen** [comment]
If you increase the maximum size in the code above, does it still get blurry? Maybe you could try 64 or something, or use the full size of the image.

I hven't tried it yet, bu maybe we can use some ideas from the following stackoverflow thread: https://stackoverflow.com/questions/18922880/html5-canvas-resize-downscale-image-high-quality/

**Comment by bijancamp** [comment]
Yes, for some reason, when I increase the maximum size with the dynamic code to 64 or the full size of the image, the icons become very blurry....

I stumbled upon that awesome Stack Overflow thread as well! Unfortunately though, since I didn't understand the mathematics presented in the solutions, I started exploring alternate solutions.... Using some of the techniques in [this article](http://devlog.disco.zone/2016/07/22/canvas-scaling/) to size and scale the canvas using the `devicePixelRatio` property, I came up with this:

```javascript
this.getDataURL = function(){
    if(_urlCache !== null)
        return _urlCache;
    
    if(_isError){
        _reloadImage();
        return IconLoader.getDefaultIcon();
    }
    
    // Some resources recommend using a ratio multiplier of 1 if the
    // devicePixelRatio is reported as less than 1...
    var devicePixelRatio = Math.max(window.devicePixelRatio || 1, 1);
        
    var size = 16;
    IconLoader._canvas.width = size * devicePixelRatio;
    IconLoader._canvas.height = size * devicePixelRatio;
        
    // Not sure if modifying the canvas's style is necessary in our case
    //IconLoader._canvas.style.width = size + "px";
    //IconLoader._canvas.style.height = size + "px";
        
    IconLoader._context.scale(devicePixelRatio, devicePixelRatio);
    
    IconLoader._context.clearRect(0, 0, size, size);
    IconLoader._context.drawImage(_img, 0, 0, size, size);
    _urlCache = IconLoader._canvas.toDataURL("image/png");
    return _urlCache;
}
```

It seems to work well on both my low-density display (devicePixelRatio of 1) and the high-density display (devicePixelRatio of 2)! On the low-density display, I tested the effects on 16x16 and 32x32 icons, and the code results in the exact same rendering as the current release (no loss in quality). On the high-density display, the code results in the exact same rendering as when the sizes were fixed to 32x32, which is a good-quality rendering.

One problem with this code is that it assumes the popup icons will have a size of 16x16 when displayed (because of the hardcoding of `16` as the minimum canvas size). If the user changes the size of the icon images to a larger size, say with the CSS rule
```css
.popup img {
    width: 32px;
    height: 32px;
}
```
then the icons will be blurry when displayed.

I think maybe ideally the minimum canvas size in the above code should be the CSS width that the icons will be displayed in. However, I am not sure how we would obtain the CSS width that will be used for the icons inside of the getDataURL method....

What do you think of this approach? 😀

**Comment by Pitmairen** [comment]
Yes, this seems like a simpler and better approach. So this is probably the way to go.   

If you increase the hardcoded size to 32 does it still look good on both the  high-density display and the low-density display?

**Comment by bijancamp** [comment]
Increasing the value of the `size` variable to 32 (in the devicePixelRatio solution) unfortunately results in blurriness on both displays whenever the icons are too small in natural size.... On the high-density display, using icons that are smaller than 64x64 results in bluriness (with `size = 32`); for the low-density display, icons smaller than 32x32 are blurry. So it seems that to have a sharp-looking icon, the size of the icon must be at least the value of the `size` variable times the devicePixelRatio. So on the low-density device, this is size 32*1 pixels; on the high-density device, this is 32*2 pixels. (This all assumes that the CSS size of the icons is 16x16.)

Additionally, it seems that the value of the `size` variable needs to be at least as great as the CSS width of the icon image. If the `size` variable is less than the CSS width, than the icon will be blurry....

Since I do not know yet how to solve the issue for users who want to use a CSS size larger than 16x16, I have made a pull request (#38) for scaling the canvas appropriately when the CSS size is the current default 16x16. 😀

**Comment by bijancamp** [comment]
This has been fixed in e21d562 (for icons that are displayed at a CSS width of 16x16).

**Comment by Pitmairen** [comment]
Thanks for the contribution, I will try to upload a new version tomorrow. 

So if the icon looks sharp when they are size*pixelRatio, what if we check the natural size of the icon before drawing and set the size variable to naturalSize/pixelRatio?

**Comment by bijancamp** [comment]
Yes, for some reason, when I set the `size` variable to `_img.naturalWidth / pixelRatio`, I don't obtain a good quality rendering for all the icons. Even the 300px Merriam-Webster SVG looks blurry, although if I change the CSS size of the icons to 75px, then the Merriam-Webster icon actually looks sharp, while the rest look blurry....

**Comment by Pitmairen** [comment]
Maybe the original icon must be a multiple of the final size for it to look good. If you resize the 300px Merriam-Webster icon to 256px, does it change anything?
